foyer-memory/src/eviction/lru.rs

```diff
@@ -0,0 +1,486 @@
+        handle.next.unwrap_unchecked().as_mut().prev = handle.prev;
+
+        handle.next = None;
+        handle.prev = None;
+    }
+}
+
+impl<K, V> Eviction for Lru<K, V>
+where
+    K: Key,
+    V: Value,
+{
+    type Handle = LruHandle<K, V>;
+    type Config = LruConfig;
+
+    unsafe fn new<S: BuildHasher>(config: &CacheConfig<Self, S>) -> Self
+    where
+        Self: Sized,
+    {
+        assert!(
+            config.eviction_config.high_priority_pool_ratio >= 0.0
+                && config.eviction_config.high_priority_pool_ratio <= 1.0,
+            "high_priority_pool_ratio_percentage must be in [0, 100], given: {}",
+            config.eviction_config.high_priority_pool_ratio
+        );
+
+        let high_priority_charges_capacity = config.capacity as f64
+            * config.eviction_config.high_priority_pool_ratio
+            / config.shards as f64;
+        let high_priority_charges_capacity = high_priority_charges_capacity as usize;
+
+        let mut head = Box::new(LruHandle::new());
+        let ptr = NonNull::new_unchecked(head.as_mut() as *mut _);
+
+        head.prev = Some(ptr);
+        head.next = Some(ptr);
+        head.in_high_priority_pool = true;
+
+        let low_priority_head = ptr;
+
+        Self {
+            head,
+            low_priority_head,
+            charges: 0,
+            high_priority_charges: 0,
+            high_priority_charges_capacity,
+            len: 0,
+        }
+    }
+
+    unsafe fn push(&mut self, mut ptr: NonNull<Self::Handle>) {
+        let handle = ptr.as_mut();
+
+        debug_assert!(handle.next.is_none());
+        debug_assert!(handle.prev.is_none());
+
+        self.charges += handle.base().charge();
+        self.len += 1;
+
+        match handle.base().context() {
+            LruContext::HighPriority => {
+                handle.in_high_priority_pool = true;
+                self.insert_ptr_after_head(ptr);
+
+                self.high_priority_charges += handle.base().charge();
+
+                if self.is_head_ptr(self.low_priority_head) {
+                    self.low_priority_head = ptr;
+                }
+
+                self.may_overflow_high_priority_pool();
+            }
+            LruContext::LowPriority => {
+                handle.in_high_priority_pool = false;
+                self.insert_ptr_after_low_priority_head(ptr);
+            }
+        }
+    }
+
+    unsafe fn pop(&mut self) -> Option<NonNull<Self::Handle>> {
+        let ptr = self.remove_ptr_before_head()?;
+        let handle = ptr.as_ref();
+
+        debug_assert!(handle.next.is_none());
+        debug_assert!(handle.prev.is_none());
+
+        if handle.in_high_priority_pool {
+            self.high_priority_charges -= handle.base.charge();
+        }
+        self.charges -= handle.base.charge();
+
+        self.len -= 1;
+
+        Some(ptr)
+    }
+
+    unsafe fn access(&mut self, _: NonNull<Self::Handle>) {}
```

Comment:
Is empty implementation intentional?

Reply:
Yep. `fn access()` should not be implemented by either FIFO or LRU. Because when the cache hits, the handle will be removed from the eviction container and referenced externally. After all external references are dropped, The entry will be inserted into the container again. So there is nothing to do with FIFO or LRU.

If the eviction policy is LFU, it should update the access frequency.

Reply:
Wait, the implementation of FIFO is wrong. Let me fix it.